### PR TITLE
(DOCSP-14311): Android Testing Guides -- unit/integration

### PR DESCRIPTION
**Pull request by nathan-contino-mongo**
### Jira

- https://jira.mongodb.org/browse/DOCSP-14311

### Staged Changes (Requires MongoDB Corp SSO)

- [Android -Testing](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/android-testing-enhancements/sdk/android/advanced-guides/testing/)

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/android/advanced-guides/testing.txt [hunk 1]
```diff
@@ -12,24 +12,311 @@ Testing - Android SDK
...
+   instance's ``countDown()`` method.
+#. At the position in your code where you need to wait for the logic,
+   in a ``try``/``catch`` block that handles an ``InterruptedException``,
+   call that ``CountDownLatch`` instance's ``await()`` method.
+#. Pass a timeout interval and unit to ``await()``, and wrap
+   the call in a ``Assert.assertTrue()`` assertion. If the logic takes
+   too long and the ``await()`` call times out, it'll return ``false``
+   and fail your test.
+
+The following example uses an ``ActivityScenario``, an empty testing
```
Comment (cbush) [inline source/sdk/android/advanced-guides/testing.txt:102]: I wonder if we should put the example up front and the explanation after? Again, as a reader, it's not clear that it will be shown how to do any of this, which feels frustrating.

File: source/sdk/android/advanced-guides/testing.txt [hunk 2]
```diff
@@ -12,24 +12,311 @@ Testing - Android SDK
...
+       .. literalinclude:: /examples/generated/android/sync/TestTest.codeblock.test-logic-looper.kt
+         :language: kotlin
+
+   .. tab::
+       :tabid: java
+
+       .. literalinclude:: /examples/generated/android/sync/TestTest.codeblock.test-logic-looper.java
+         :language: java
+
+Full Example
```
Comment (cbush) [inline source/sdk/android/advanced-guides/testing.txt:153]: I expected "A way to delay unit tests execution" here.

File: source/sdk/android/advanced-guides/testing.txt [hunk 3]
```diff
@@ -12,24 +12,311 @@ Testing - Android SDK
...
+         :language: java
 
 .. seealso::
 
-   See the `Unit Testing Example App
-   <https://github.com/realm/realm-java/tree/master/examples/unitTestExample>`__
-   for an example of how {+service-short+} can be combined with JUnit3,
-   JUnit4, Robolectric, Mockito and PowerMock.
+   See the :github:`{+service-short+} Documentation Examples App
+   <mongodb/docs-realm/tree/master/examples/android>` for an example
```
Comment (cbush) [inline source/sdk/android/advanced-guides/testing.txt:176]: I like this!

**Review by cbush — APPROVED** [review]
Great work!

File: source/sdk/android/advanced-guides/testing.txt [hunk 4]
```diff
@@ -12,24 +12,310 @@ Testing - Android SDK
    :depth: 2
    :class: singlecol
 
-Unit Testing
-------------
+You can test your application using unit tests or integration tests.
+*Unit tests* only assess the logic written in your application's code.
+*Integration tests* assess your application logic, database queries and
+writes, and calls to your application's backend (if you use
```
[the remark this answers is not in the document]
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:18]: Good point. Revised this to "if you have one" because I'm trying not to assume that all users are using sync :)

File: source/sdk/android/advanced-guides/testing.txt [hunk 5]
```diff
@@ -12,24 +12,310 @@ Testing - Android SDK
...
+Because the SDK uses C++ code via Android Native for data
+storage, unit testing requires you to entirely mock interactions with
+{+client-database+}. Prefer integration tests for logic that requires
+extensive interaction with the database.
+
+Integration Tests
+-----------------
+
+To integration test an application that uses the {+service-short+} SDK,
+you'll need:
```
[the remark this answers is not in the document]
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:50]: Totally reworded this to describe how this section walks you through these tasks. Thanks!

File: source/sdk/android/advanced-guides/testing.txt (hunk 1 above)
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:102]: Hm, definitely a fair concern. I worry that readers might glance over the example without understanding what's going on in it (since it looks relatively simple), whereas if there's a bunch of explanation first it's clear that the explanation explains that code block. Since readers don't like scrolling backwards on a page, this preserves that "forward flow" of moving through the docs, instead of start-stop-backpedal-frontpedal-start-stop.

I'll at least add mention of examples above so users get the expectation of explanation-example if they read the intro to this section.

File: source/sdk/android/advanced-guides/testing.txt (hunk 2 above)
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:153]: Added a section on this that separately explains CountDownLatch (a built-in alternative to our Expectation class, which was a tad too much bespoke code to toss in this guide).

File: examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/kotlin/TestTest.kt [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import org.junit.Assert
+import org.junit.Test
+
+class TestTest {
+
+    @Test
+    fun testTesting() {
+        // :code-block-start: initialize-realm-with-context
+        var testActivity: Activity? = null
+        val scenario: ActivityScenario<BasicActivity>? =
```
Comment (dacharyc) [inline examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/kotlin/TestTest.kt:27]: Same as the Java example - I'd love to see something more concrete here as an example vs. `BasicActivity`.

File: source/sdk/android/advanced-guides/testing.txt [hunk 7]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+{+client-database+} functionality such as 
+:ref:`Live objects <android-live-object>` and change notifications only
+work on :android:`Looper <reference/android/os/Looper>` threads.
+Threads configured with a ``Looper`` object pass events over a message
+loop coordinated by the ``Looper``. Test functions normally don't have
+a ``Looper`` object, and configuring one to work in your tests can be
+very error-prone. Instead, you can use the :android:`Activity.runOnUiThread()
+<reference/android/app/Activity#runOnUiThread(java.lang.Runnable)>`
+method of your test activity to execute logic on a thread that already
+has a ``Looper`` configured. Combine ``Activity.runOnUiThread()`` with
```
Comment (dacharyc) [inline source/sdk/android/advanced-guides/testing.txt:119]: Nit: this is a long paragraph that is dense. I'd recommend splitting this into two paragraphs, either at the "Instead..." or "Combine..."

File: source/sdk/android/advanced-guides/testing.txt [hunk 8]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+a ``CountDownLatch`` as described in the :ref:`next section
+<android-testing-delay-while-async>` above to prevent your test from completing
```
Comment (dacharyc) [inline source/sdk/android/advanced-guides/testing.txt:121]: I'm confused by this "in the next section above" - I think you might be referring to the ``CountDownLatch`` you mentioned in the example above, or the next section where you describe delay test execution, but as written I'm not sure what we're talking about.

File: source/sdk/android/advanced-guides/testing.txt [hunk 9]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+to wait for those async calls to complete. Otherwise, your tests will
+exit before your asynchronous (or multi-threaded) calls run. This example
+uses Java's built-in `CountDownLatch
+<https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CountDownLatch.html>`__. Follow these steps to use a ``CountDownLatch`` in your own tests:
+
+1. Instantiate a ``CountDownLatch`` with a count of 1.
+#. After running the async logic your test needs to wait for, call that
+   ``CountDownLatch`` instance's ``countDown()`` method.
+#. At the position in your code where you need to wait for the logic,
+   in a ``try``/``catch`` block that handles an ``InterruptedException``,
```
Comment (dacharyc) [inline source/sdk/android/advanced-guides/testing.txt:156]: This is a lot of prepositions and makes this bullet difficult to follow. Maybe reword to something like:
> In your code, where you need to wait for the logic, add a ``try``/``catch`` block that handles...

File: source/sdk/android/advanced-guides/testing.txt [hunk 10]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+   call that ``CountDownLatch`` instance's ``await()`` method.
+#. Pass a timeout interval and unit to ``await()``, and wrap
+   the call in a ``Assert.assertTrue()`` assertion. If the logic takes
+   too long and the ``await()`` call times out, it'll return ``false``
```
Comment (dacharyc) [inline source/sdk/android/advanced-guides/testing.txt:160]: In "it'll return ``false``" - it's not obvious to me what "it" is in this case. Maybe better to say something like "If the logic takes too long, the ``await()`` call times out, returning ``false`` and failing your test.

File: source/sdk/android/advanced-guides/testing.txt [hunk 11]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+   .. tab::
+       :tabid: java
+
+       .. literalinclude:: /examples/generated/android/sync/TestTest.codeblock.wait-for-async.java
+         :language: java
+
+Full Example
+~~~~~~~~~~~~
+
+The following example shows a full Junit instrumented ``androidTest``
```
Comment (dacharyc) [inline source/sdk/android/advanced-guides/testing.txt:184]: I don't think we mention `Junit` before here. Might you want to link to whatever Junit is?

Hmm, just Googled - it looks like this is typically written JUnit. I see stuff for 3, 4, and 5 - might we want to explain why we're choosing JUnit, and what version we're using?

File: source/sdk/android/advanced-guides/testing.txt (hunk 3 above)
Comment (dacharyc) [reply source/sdk/android/advanced-guides/testing.txt:204]: Same! ❤️

File: source/sdk/android/advanced-guides/testing.txt [hunk 12]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+         :language: java
+
+After mocking a {+realm+}, you'll have to configure data for your test
+cases. See the full example below for some examples of how you can
+provide testing data in unit tests.
+
+Full Example
+~~~~~~~~~~~~
+
+The following example shows a full Junit ``test``
```
Comment (dacharyc) [inline source/sdk/android/advanced-guides/testing.txt:320]: JUnit?

**Review by dacharyc — APPROVED** [review]
Wow, this is a big lift! Great work! Left some questions/comments/suggestions, but I think it's great that you're adding this.

And I think you've put it in the right place, unfortunately.

File: source/sdk/android/advanced-guides/testing.txt (hunk 11 above)
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:184]: Good catch on the spelling -- Fixed that! As for the "why"... it basically boils down to "because android/google said so". Android doesn't really offer you much choice in this vein so I think I'll leave it as-is. Most developers never even consider an alternative because JUnit is spoonfed so much to android.

File: source/sdk/android/advanced-guides/testing.txt (hunk 8 above)
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:121]: the "above" was accidentally left in during a rewrite, thanks!

Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:121]: Also rephrased to "delay section" so we avoid ordering snafus in the future.

File: examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/kotlin/TestTest.kt (hunk 6 above)
Comment (nathan-contino-mongo) [reply examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/kotlin/TestTest.kt:27]: Most Android tests boil down to "instantiate an activity, call some methods". This example just streamlines that into "instantiate an activity, run some code *in the activity's thread using a lambda*" to make things a bit more readable. I'll keep an eye on this as feedback comes in and maybe move the logic to the activity instead if folks seem confused. For what it's worth:

- most android developers should have a pretty good notion of what an empty activity is (it's just a class that extends `Activity`)
- the unit testing example demonstrates calling logic through methods in the activity itself, so our bases are... sort of covered.

Definitely one to keep an eye on.

File: examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/java/TestTest.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+import static com.mongodb.realm.examples.RealmTestKt.YOUR_APP_ID;
+import static com.mongodb.realm.examples.RealmTestKt.getRandomPartition;
+
+public class TestTest {
+    @Test
+    public void testTesting() {
+        // :code-block-start: initialize-realm-with-context
+        AtomicReference<Activity> testActivity = new AtomicReference<Activity>();
+        ActivityScenario<BasicActivity> scenario = ActivityScenario.launch(BasicActivity.class);
```
[the remark this answers is not in the document]
Comment (nathan-contino-mongo) [reply examples/android/sync/app/src/androidTest/java/com/mongodb/realm/examples/java/TestTest.java:34]: See below. TL;DR: this is mostly so I don't have to present yet another giant entire-class code snippet to users to show the logic of the example.

File: source/sdk/android/advanced-guides/testing.txt [hunk 14]
```diff
@@ -12,24 +12,340 @@ Testing - Android SDK
...
+
+This section shows how to integration test an application that uses
+the {+service-short+} SDK. It covers the following concepts in the test
+environment:
+
+- acquiring an application context
+- executing logic on a ``Looper`` thread
+- how to delay test execution while asynchronous method calls complete
+
+Applications that use {+sync+} or a backend {+app+} also require (not
```
[the remark this answers is not in the document]
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:56]: Ha, I actually originally had a section about this! Maybe this is the wakeup call I need to reinstate it, at least as a bunch of bullet points.

[the remark this answers is not in the document]
Comment (nathan-contino-mongo) [reply source/sdk/android/advanced-guides/testing.txt:56]: Added a section on separate apps as well as a section on separate clusters with a link to relevant information each, as well as some justification of why this is important. Keeping it lean for now but I think it addresses this!

**Review by clementetb — APPROVED** [review]
LGTM
